rqt_py_common/src/rqt_py_common/topic_helpers.py

```diff
@@ -28,161 +28,167 @@
 # ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 
-import roslib
-import roslib.msgs
-import roslib.message
-from rostopic import get_topic_type
-from python_qt_binding.QtCore import qDebug
+# Author: Michael Lautman
 
+from ament_index_python import get_resource
 
-def get_type_class(type_name):
-    if roslib.msgs.is_valid_constant_type(type_name):
-        if type_name == 'string':
-            return str
-        elif type_name == 'bool':
-            return bool
-        else:
-            return type(roslib.msgs._convert_val(type_name, 0))
-    else:
-        return roslib.message.get_message_class(type_name)
+from rclpy import logging
 
 
-def get_field_type(topic_name):
+def get_topic_names_and_types(node=None):
     """
-    Get the Python type of a specific field in the given registered topic.
-    If the field is an array, the type of the array's values are returned and the is_array flag is set to True.
-    This is a static type check, so it works for unpublished topics and with empty arrays.
+    Get avaliable topic names and types.
 
-    :param topic_name: name of field of a registered topic, ``str``, i.e. '/rosout/file'
-    :returns: field_type, is_array
+    Note: in ROS2, only nodes can query the topic information
+    @param node: a ROS node
+    @type node: rclpy.node.Node or None
+
+    If node is None, then this method will create a node, use it to get topic
+    information and then destroy the node.
     """
-    # get topic_type and message_evaluator
-    topic_type, real_topic_name, _ = get_topic_type(topic_name)
-    if topic_type is None:
-        # qDebug('topic_helpers.get_field_type(%s): get_topic_type failed' % (topic_name))
-        return None, False
+    if node is not None:
+        return node.get_topic_names_and_types()
 
-    message_class = roslib.message.get_message_class(topic_type)
-    if message_class is None:
-        qDebug('topic_helpers.get_field_type(%s): get_message_class(%s) failed' %
-               (topic_name, topic_type))
-        return None, False
+    import rclpy
+    shutdown_rclpy = False
+    if not rclpy.ok():
+        shutdown_rclpy = True
+        rclpy.init()
 
-    slot_path = topic_name[len(real_topic_name):]
-    return get_slot_type(message_class, slot_path)
+    node = rclpy.create_node("TopicHelpers__get_topic_names_and_types")
 
+    # Give the node time to learn about the graph
+    rclpy.spin_once(node, timeout_sec=0.05)
```

Comment:
This looks pretty fragile. This timeout is very likely not enough - and no hard coded value will be good for all cases.

Reply:
Is there a better way to do this? I modeled this after https://github.com/ros2/ros2cli/blob/master/ros2cli/ros2cli/node/direct.py

Reply:
Addressed in #155

Reply:
@dirk-thomas I have a separate PR for using the daemon approach that you detailed. Also, @brawner's PR https://github.com/ros-visualization/rqt/pull/161